File: src/main/kotlin/com/chopeks/SquirrelBundle.java
package com.chopeks;

import com.intellij.AbstractBundle;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.PropertyKey;

import java.lang.ref.Reference;
import java.lang.ref.SoftReference;
import java.util.ResourceBundle;

public class SquirrelBundle {

  public static String message(@NotNull @PropertyKey(resourceBundle = BUNDLE) String key, @NotNull Object... params) {
    return AbstractBundle.message(getBundle(), key, params);
  }

  private static Reference<ResourceBundle> ourBundle;
  @NonNls public static final String BUNDLE = "com.sqide.SquirrelBundle";

  private SquirrelBundle() {
  }

  private static ResourceBundle getBundle() {
    ResourceBundle bundle = com.intellij.reference.SoftReference.dereference(ourBundle);
    if (bundle == null) {
      bundle = ResourceBundle.getBundle(BUNDLE);
      ourBundle = new SoftReference<ResourceBundle>(bundle);
    }
    return bundle;
  }
}

File: src/main/kotlin/com/chopeks/SquirrelBundle.kt
package com.chopeks

import org.jetbrains.annotations.NonNls
import org.jetbrains.annotations.PropertyKey
import java.util.*

object SquirrelBundle {
	private const val BUNDLE: @NonNls String = "com.chopeks.SquirrelBundle"
	private var ourBundle: ResourceBundle? = null

	@JvmStatic
	fun message(key: @PropertyKey(resourceBundle = BUNDLE) String, vararg params: Any?): String {
		return try {
			String.format(bundle!!.getString(key), *params)
		} catch (e: MissingResourceException) {
			key
		}
	}

	private val bundle: ResourceBundle?
		get() {
			if (ourBundle == null) {
				ourBundle = try {
					ResourceBundle.getBundle(BUNDLE, Locale.getDefault())
				} catch (e: MissingResourceException) {
					ResourceBundle.getBundle(BUNDLE, Locale.ENGLISH)
				}
			}
			return ourBundle
		}
}
